packages/react-router/lib/router/router.ts

```diff
@@ -4617,36 +4884,214 @@ async function loadLazyRouteModule(
 }
 
 // Default implementation of `dataStrategy` which fetches all loaders in parallel
-async function defaultDataStrategy({
-  matches,
-}: DataStrategyFunctionArgs): ReturnType<DataStrategyFunction> {
-  let matchesToLoad = matches.filter((m) => m.shouldLoad);
+async function defaultDataStrategy(
+  args: DataStrategyFunctionArgs<unknown>
+): ReturnType<DataStrategyFunction<unknown>> {
+  let matchesToLoad = args.matches.filter((m) => m.shouldLoad);
+  let keyedResults: Record<string, DataStrategyResult> = {};
   let results = await Promise.all(matchesToLoad.map((m) => m.resolve()));
-  return results.reduce(
-    (acc, result, i) =>
-      Object.assign(acc, { [matchesToLoad[i].route.id]: result }),
-    {}
+  results.forEach((result, i) => {
+    keyedResults[matchesToLoad[i].route.id] = result;
+  });
+  return keyedResults;
+}
+
+// Middleware-enabled implementation of `dataStrategy` which calls middleware
+// and fetches all loaders in parallel
+async function defaultDataStrategyWithMiddleware(
+  args: DataStrategyFunctionArgs<unknown>
+): ReturnType<DataStrategyFunction<unknown>> {
+  // Short circuit all the middleware logic if we have no middlewares
+  if (!args.matches.some((m) => m.route.unstable_middleware)) {
+    return defaultDataStrategy(args);
+  }
+
+  // Determine how far down we'll be loading so we only run middleware to that
+  // point.  This prevents us from calling middleware below an action error
+  // boundary below which we don't run loaders
+  let lastIndex = args.matches.length - 1;
+  for (let i = lastIndex; i >= 0; i--) {
+    if (args.matches[i].shouldLoad) {
+      lastIndex = i;
+      break;
+    }
+  }
+
+  let results = await runMiddlewarePipeline(
+    args,
+    lastIndex,
+    false,
+    async (keyedResults: Record<string, DataStrategyResult>) => {
+      Object.assign(keyedResults, await defaultDataStrategy(args));
+    },
+    (e, keyedResults) => {
+      // we caught an error running the middleware, copy that overtop any
+      // non-error result for the route
+      Object.assign(keyedResults, {
+        [e.routeId]: { type: "error", result: e.error },
+      });
+    }
   );
+  return results as Record<string, DataStrategyResult>;
+}
+
+type MutableMiddlewareState = {
+  keyedResults: Record<string, DataStrategyResult>;
+  propagateResult: boolean;
+};
+
+export async function runMiddlewarePipeline(
```

Comment:
This is the main implementation to run the middlewares and provide the `next` method